File: frontend/src/services/stressAnalysis.ts
/**
 * Comprehensive stress analysis service for communicating with the backend
 * stress analysis API that uses spaCy, CMU dictionary, and G2P fallback.
 */

export interface WordAnalysis {
  word: string
  pos: string  // Part of speech tag from spaCy
  syllables: string[]
  stress_pattern: number[]  // 0=unstressed, 1=primary, 2=secondary
  reasoning: string  // Explanation of how stress was determined
  char_positions: number[]  // Character positions for stress mark placement
  confidence: number
}

export interface StressAnalysisResult {
  text: string
  total_syllables: number
  stressed_syllables: number
  processing_time_ms: number
  words: WordAnalysis[]
}

export interface BatchStressAnalysisResult {
  total_lines: number
  total_processing_time_ms: number
  lines: Array<{
    line_number: number
    text: string
    total_syllables: number
    stressed_syllables: number
    processing_time_ms: number
    words: WordAnalysis[]
  }>
}

export interface AnalyzerStatus {
  status: 'ready' | 'error'
  message: string
  components: {
    spacy_model?: string
    g2p_model?: string
    cmu_dictionary?: string
    total_words?: number
    cache_size?: number
    cache_hits?: number
    cache_misses?: number
  }
}

class ComprehensiveStressAnalysisService {
  private baseURL: string

  constructor() {
    this.baseURL = import.meta.env.VITE_API_URL || 'http://localhost:8001'
  }

  /**
   * Analyze stress patterns for a single text using the comprehensive algorithm
   */
  async analyzeText(
    text: string,
    context: 'lyrical' | 'conversational' = 'lyrical'
  ): Promise<StressAnalysisResult> {
    const response = await fetch(`${this.baseURL}/api/stress/analyze`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        text,
        context
      }),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || `HTTP ${response.status}: ${response.statusText}`)
    }

    return response.json()
  }

  /**
   * Analyze stress patterns for multiple lines of text in batch
   */
  async analyzeBatch(
    lines: string[],
    context: 'lyrical' | 'conversational' = 'lyrical'
  ): Promise<BatchStressAnalysisResult> {
    const response = await fetch(`${this.baseURL}/api/stress/analyze-batch`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        lines,
        context
      }),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || `HTTP ${response.status}: ${response.statusText}`)
    }

    return response.json()
  }

  /**
   * Get status of the comprehensive stress analyzer components
   */
  async getAnalyzerStatus(): Promise<AnalyzerStatus> {
    const response = await fetch(`${this.baseURL}/api/stress/analyzer-status`)

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`)
    }

    return response.json()
  }

  /**
   * Extract lines from Lexical JSON structure for batch analysis
   */
  extractLinesFromLexicalJSON(lexicalJSON: unknown): string[] {
    const lines: string[] = []

    function traverseNodes(nodes: unknown[]) {
      for (const node of nodes) {
        const nodeObj = node as Record<string, unknown>
        if (nodeObj.type === 'section-paragraph' || nodeObj.type === 'paragraph') {
          // Extract text content from paragraph node
          const textContent = this.extractTextFromNode(nodeObj)
          if (textContent.trim()) {
            lines.push(textContent.trim())
          }
        }

        // Recursively traverse children
        if (nodeObj.children && Array.isArray(nodeObj.children)) {
          traverseNodes(nodeObj.children as unknown[])
        }
      }
    }

    const lexicalObj = lexicalJSON as Record<string, unknown>
    const root = lexicalObj?.root as Record<string, unknown>
    if (root?.children && Array.isArray(root.children)) {
      traverseNodes.call(this, root.children as unknown[])
    }

    return lines
  }

  /**
   * Extract text content from a Lexical node
   */
  private extractTextFromNode(node: unknown): string {
    if (!node || typeof node !== 'object') return ''

    const nodeObj = node as Record<string, unknown>

    if (nodeObj.type === 'text') {
      return String(nodeObj.text || '')
    }

    if (nodeObj.children && Array.isArray(nodeObj.children)) {
      return nodeObj.children
        .map((child: unknown) => this.extractTextFromNode(child))
        .join('')
    }

    return ''
  }

  /**
   * Convert stress analysis results back to character positions for UI rendering
   */
  getStressMarksForText(text: string, words: WordAnalysis[]): Array<{
    position: number
    stress: number
    word: string
    syllable: string
  }> {
    const stressMarks: Array<{
      position: number
      stress: number
      word: string
      syllable: string
    }> = []

    let textPosition = 0

    for (const wordAnalysis of words) {
      // Find word position in text
      const wordStartIndex = text.indexOf(wordAnalysis.word, textPosition)
      if (wordStartIndex === -1) continue

      // Map stress marks to character positions
      for (let i = 0; i < wordAnalysis.char_positions.length; i++) {
        const charPos = wordAnalysis.char_positions[i]
        const stressLevel = wordAnalysis.stress_pattern[i] || 0
        const syllable = wordAnalysis.syllables[i] || ''

        stressMarks.push({
          position: wordStartIndex + charPos,
          stress: stressLevel,
          word: wordAnalysis.word,
          syllable
        })
      }

      textPosition = wordStartIndex + wordAnalysis.word.length
    }

    return stressMarks
  }

  /**
   * Check if the comprehensive analyzer is available and ready
   */
  async isReady(): Promise<boolean> {
    try {
      const status = await this.getAnalyzerStatus()
      return status.status === 'ready'
    } catch {
      return false
    }
  }
}

// Export singleton instance
export const comprehensiveStressAnalysis = new ComprehensiveStressAnalysisService()
export default comprehensiveStressAnalysis

File: backend/app/stress_analysis.py
"""
Comprehensive stress analysis service for lyrics based on linguistic principles.

Implements the robust algorithm specified in .private_docs/Lyric_Stress_Algorithm.md
using spaCy POS tagging, CMU dictionary lookup, and G2P fallback for comprehensive
stress detection that respects English prosody.
"""

import re
from dataclasses import dataclass
from functools import lru_cache
from typing import List, Optional, Tuple

# Import NLP libraries
import spacy
from g2p_en import G2p

# Import existing dictionary service
from .dictionary import get_cmu_dictionary


@dataclass
class WordAnalysis:
    """Analysis result for a single word."""

    word: str
    pos: str  # Part of speech tag from spaCy
    syllables: List[str]
    stress_pattern: List[int]  # 0=unstressed, 1=primary, 2=secondary
    reasoning: str  # Explanation of how stress was determined
    char_positions: List[int]  # Character positions for stress mark placement
    confidence: float = 1.0


@dataclass
class StressAnalysisResult:
    """Complete stress analysis result for text input."""

    text: str
    words: List[WordAnalysis]
    total_syllables: int
    stressed_syllables: int
    processing_time_ms: float


class ComprehensiveStressAnalyzer:
    """
    Comprehensive stress analyzer implementing the robust algorithm from
    the Lyric_Stress_Algorithm.md specification.
    """

    def __init__(self):
        # Initialize NLP components
        self.nlp = self._load_spacy_model()
        self.g2p = G2p()
        self.cmu_dict = get_cmu_dictionary()

        # Vowel sets for phoneme and orthographic analysis
        self.VOWELS = {
            "AA",
            "AE",
            "AH",
            "AO",
            "AW",
            "AY",
            "EH",
            "ER",
            "EY",
            "IH",
            "IY",
            "OW",
            "OY",
            "UH",
            "UW",
        }
        self.ORTHO_VOWELS = set("aeiouyAEIOUY")

        # Contraction patterns
        self.CONTRACTIONS = {
            "n't": ["not"],  # don't, can't, won't
            "'re": ["are"],  # you're, they're
            "'ve": ["have"],  # I've, we've
            "'ll": ["will"],  # I'll, you'll
            "'d": ["would", "had"],  # I'd, you'd (context-dependent)
            "'s": ["is", "has"],  # it's, he's (context-dependent)
        }

    def _load_spacy_model(self) -> spacy.Language:
        """Load spaCy English model with error handling."""
        from .nlp_setup import ensure_spacy_model_installed

        model_name = "en_core_web_sm"

        # Ensure model is installed (will attempt download if missing)
        if not ensure_spacy_model_installed(model_name):
            raise RuntimeError(
                f"spaCy model '{model_name}' could not be loaded or downloaded. "
                f"For production deployments, ensure the Docker image includes: "
                f"RUN python -m spacy download {model_name}"
            )

        try:
            return spacy.load(model_name)
        except OSError as e:
            raise RuntimeError(
                f"spaCy model '{model_name}' failed to load after installation: {e}"
            )

    @lru_cache(maxsize=1000)
    def get_phonemes(self, word: str) -> str:
        """Get phonemes for a word using CMU dict or G2P fallback."""
        # First try CMU dictionary
        cmu_result = self.cmu_dict.lookup(word.lower())
        if cmu_result:
            return " ".join(cmu_result.phonemes)

        # Fallback to G2P
        g2p_result = self.g2p(word)
        # Filter to get ARPAbet-style phonemes
        phonemes = [p for p in g2p_result if re.match(r"[A-Z]+[0-2]?$", p)]
        return " ".join(phonemes)

    def extract_stress_digits(self, phonemes_str: str) -> List[int]:
        """Extract stress pattern from phoneme string."""
        stress_digits = []
        for phoneme in phonemes_str.split():
            match = re.match(r"([A-Z]+)([0-2]?)", phoneme)
            if not match:
                continue
            base, stress_digit = match.groups()
            if base in self.VOWELS:
                stress_digits.append(int(stress_digit) if stress_digit else 0)
        return stress_digits

    def is_content_monosyllable(self, token) -> bool:
        """Check if a token is a content word (typically stressed)."""
        return token.pos_ in {"NOUN", "VERB", "ADJ", "ADV", "INTJ"}

    def analyze_monosyllable_stress(self, token) -> Tuple[int, str]:
        """
        Analyze stress for monosyllabic words using POS-based rules.
        Returns (stress_level, reasoning).
        """
        word = token.lower_
        pos = token.pos_

        # Special cases first
        if word == "not" or token.text.endswith("n't"):
            return 1, "negation_stressed"

        if word == "there":
            # Check for existential "there is/are" pattern
            next_token = token.nbor(1) if token.i + 1 < len(token.doc) else None
            if (
                next_token
                and next_token.lemma_ == "be"
                and next_token.pos_ in {"AUX", "VERB"}
            ):
                return 0, "existential_there_unstressed"
            return 1, "locative_there_stressed"

        # Phrasal verb particles are stressed
        if token.dep_ == "prt":
            return 1, "phrasal_verb_particle_stressed"

        # Function vs content word classification
        if pos in {"DET", "PRON", "ADP", "AUX", "PART", "CCONJ", "SCONJ"}:
            return 0, f"function_word_unstressed_{pos.lower()}"
        elif pos in {"NOUN", "VERB", "ADJ", "ADV", "INTJ"}:
            return 1, f"content_word_stressed_{pos.lower()}"
        else:
            # Default to unstressed for uncertain cases
            return 0, f"default_unstressed_{pos.lower()}"

    def map_syllables_to_chars(self, word: str, n_syllables: int) -> List[int]:
        """
        Map phonetic syllables to character positions in orthography.
        Returns list of character indices where stress marks should be placed.
        """
        # Find vowel positions in the word
        vowel_positions = [i for i, char in enumerate(word) if char.lower() in "aeiouy"]

        if not vowel_positions:
            return []

        # If we have the right number of vowels, map directly
        if len(vowel_positions) == n_syllables:
            return vowel_positions

        # Handle common mismatches
        if len(vowel_positions) > n_syllables:
            # Too many vowels - likely silent 'e' or diphthongs
            # Keep first n_syllables vowels, but check for silent 'e'
            if word.endswith("e") and len(vowel_positions) > 1:
                # Remove final 'e' if it's likely silent
                vowel_positions = vowel_positions[:-1]

            # Still too many? Take proportional spacing
            if len(vowel_positions) > n_syllables:
                step = len(vowel_positions) / n_syllables
                return [vowel_positions[int(i * step)] for i in range(n_syllables)]

        elif len(vowel_positions) < n_syllables:
            # Too few vowels - extend the last position
            while len(vowel_positions) < n_syllables:
                vowel_positions.append(vowel_positions[-1])

        return vowel_positions[:n_syllables]

    def analyze_text(self, text: str, context: str = "lyrical") -> StressAnalysisResult:
        """
        Perform comprehensive stress analysis on input text.

        Args:
            text: The text to analyze
            context: Context hint ("lyrical" vs "conversational") for prosody rules

        Returns:
            Complete stress analysis result
        """
        import time

        start_time = time.time()

        # Process text with spaCy
        doc = self.nlp(text)

        word_analyses: List[WordAnalysis] = []
        total_syllables = 0
        stressed_syllables = 0

        for token in doc:
            # Skip punctuation and whitespace
            if not token.text.strip() or token.is_punct:
                continue

            word = token.text
            pos = token.pos_

            # Get phonemes and syllable count
            phonemes = self.get_phonemes(word)
            stress_pattern = self.extract_stress_digits(phonemes)
            n_syllables = len(stress_pattern) if stress_pattern else 1

            # Determine stress pattern based on syllable count
            if n_syllables == 1:
                # Monosyllabic - use POS-based rules
                stress_level, reasoning = self.analyze_monosyllable_stress(token)
                stress_pattern = [stress_level]
                syllables = [word]

            elif stress_pattern:
                # Multisyllabic with CMU data
                cmu_result = self.cmu_dict.lookup(word.lower())
                if cmu_result:
                    syllables = cmu_result.syllables
                    reasoning = "cmu_dictionary_multisyllable"
                else:
                    # Use G2P result
                    syllables = self._approximate_syllables(word, n_syllables)
                    reasoning = "g2p_fallback_multisyllable"

            else:
                # Fallback - treat as single stressed syllable
                stress_pattern = [1]
                syllables = [word]
                reasoning = "fallback_single_stressed"

            # Map to character positions
            char_positions = self.map_syllables_to_chars(word, len(stress_pattern))

            # Create analysis result
            analysis = WordAnalysis(
                word=word,
                pos=pos,
                syllables=syllables,
                stress_pattern=stress_pattern,
                reasoning=reasoning,
                char_positions=char_positions,
                confidence=0.95 if reasoning.startswith("cmu_") else 0.8,
            )

            word_analyses.append(analysis)
            total_syllables += len(stress_pattern)
            stressed_syllables += sum(1 for s in stress_pattern if s > 0)

        # Apply prosodic smoothing if enabled
        if len(word_analyses) > 2:
            self._apply_prosodic_smoothing(word_analyses)

        processing_time = (time.time() - start_time) * 1000

        return StressAnalysisResult(
            text=text,
            words=word_analyses,
            total_syllables=total_syllables,
            stressed_syllables=stressed_syllables,
            processing_time_ms=processing_time,
        )

    def _approximate_syllables(self, word: str, n_syllables: int) -> List[str]:
        """Approximate syllable boundaries for a word."""
        if n_syllables <= 1:
            return [word]

        # Find vowel positions for splitting
        vowel_positions = [i for i, char in enumerate(word.lower()) if char in "aeiouy"]

        if len(vowel_positions) >= n_syllables:
            # Split at consonant clusters between vowels
            syllables = []
            # Split at consonant clusters between vowels (word_len available if needed)

            for i in range(n_syllables):
                if i == 0:
                    # First syllable
                    if len(vowel_positions) > 1:
                        split_point = min(
                            vowel_positions[1],
                            (vowel_positions[0] + vowel_positions[1]) // 2 + 1,
                        )
                        syllables.append(word[:split_point])
                    else:
                        syllables.append(word)
                elif i == n_syllables - 1:
                    # Last syllable
                    prev_split = len("".join(syllables))
                    syllables.append(word[prev_split:])
                else:
                    # Middle syllables
                    prev_split = len("".join(syllables))
                    next_vowel_idx = min(i + 1, len(vowel_positions) - 1)
                    split_point = min(
                        vowel_positions[next_vowel_idx],
                        (vowel_positions[i] + vowel_positions[next_vowel_idx]) // 2 + 1,
                    )
                    syllables.append(word[prev_split:split_point])

            return [s for s in syllables if s]

        # Fallback to simple division
        syllable_length = len(word) / n_syllables
        syllables = []
        for i in range(n_syllables):
            start = int(i * syllable_length)
            end = int((i + 1) * syllable_length) if i < n_syllables - 1 else len(word)
            syllables.append(word[start:end])

        return syllables

    def _apply_prosodic_smoothing(self, word_analyses: List[WordAnalysis]) -> None:
        """
        Apply optional prosodic smoothing to avoid awkward stress clusters.
        Modifies word_analyses in place.
        """
        # Look for 3+ consecutive stressed syllables
        stress_sequence = []
        for analysis in word_analyses:
            stress_sequence.extend(analysis.stress_pattern)

        # Find clusters of 3+ stressed syllables
        i = 0
        while i < len(stress_sequence) - 2:
            if (
                stress_sequence[i] > 0
                and stress_sequence[i + 1] > 0
                and stress_sequence[i + 2] > 0
            ):

                # Found a stress cluster - demote the middle one if it's a function word
                # This is a simplified implementation; full version would track word boundaries
                if stress_sequence[i + 1] == 1:  # Only demote primary stress
                    stress_sequence[i + 1] = 0

                i += 3  # Skip past this cluster
            else:
                i += 1

        # Map modified stress sequence back to word analyses
        # This is a simplified implementation - production would need careful boundary tracking


# Global analyzer instance
_stress_analyzer: Optional[ComprehensiveStressAnalyzer] = None


def get_stress_analyzer() -> ComprehensiveStressAnalyzer:
    """Get or create the global stress analyzer instance."""
    global _stress_analyzer
    if _stress_analyzer is None:
        _stress_analyzer = ComprehensiveStressAnalyzer()
    return _stress_analyzer


def analyze_stress(text: str, context: str = "lyrical") -> StressAnalysisResult:
    """Convenience function for stress analysis."""
    return get_stress_analyzer().analyze_text(text, context)
